src/stdune/io.ml

```diff
@@ -17,6 +17,38 @@ let input_lines =
   in
   fun ic -> loop ic []
 
+let input_zero_from_buffer from buf =
+  match Bytes.index_from_opt buf from '\x00' with
+  | None -> (None, from)
+  | Some eos ->
+    (Some (Bytes.sub_string buf ~pos:from ~len:(eos - from)), eos + 1)
+
+let input_zero_separated =
+  let buf_size = 65536 in
+  let buf = Bytes.create buf_size in
+  let rec process_buf buf pos acc =
+    match input_zero_from_buffer pos buf with
+    | Some s, eos -> process_buf buf eos (s :: acc)
+    | None, rest -> (acc, rest)
+  in
+  let rec loop ic pos acc =
```

Comment:
This code looks buggy. If one of the strings has length `>= Bytes.length buf`, this code will keep reading the string until `buf` gets filled up (ie `pos = buf_size`) and then the exit clause will trigger `res = 0` prematurely.

Either we need to fail if we come across a long string or you need to keep some kind of variable-length buffer to store the string that is currently being read.

Reply:
Oh indeed the code doesn't support a stringer longer than `>= Bytes.length buf`; that's OK for filenames but not in general.

As I mentioned I hacked something quick and dirty, but surely there has to be a reference implementation we could use [maybe in `Base`?] instead of rolling our own, right? This kind of stuff can be tricky (specially perf-wise) so I prefer to use a tested implementation.

Otherwise I'd be happy to do a pass on the code a fix the bugs and maybe tweak a few things here.

Reply:
I think it is OK to be simple here. You can look at the implementation of `input_line` for inspiration; there they keep a list of fixed-size buffers instead of resizing a single buffer to fit the string being read.
https://github.com/ocaml/ocaml/blob/10364b67d5d7af71b8626c805ffc0ec49d9b1522/stdlib/stdlib.ml#L439-L465

Reply:
They do rely on `caml_input_scan_line` tho, we don't have anything equivalent so I'm not sure how we can simplify.

Reply:
Ok, I did a new, I hope simpler and more correct version, can you have a look while I add the test?